fe1-web/model/network/method/message/Message.ts

```diff
@@ -111,68 +112,57 @@ export class Message {
   }
 
   /**
-   * Creates a Message object from a given MessageData and signatures
-   * We don't add the channel property here as we don't want to send that over the network
-   * It signs the messages with the most recent pop token if it exists. Otherwise it uses the
-   * public key
-   *
-   * @param data The MessageData to be signed and hashed
-   * @param witnessSignatures The signatures of the witnesses
+   * Creates a Message object from a given MessageData and signatures.
+   * We don't add the channel property here as we don't want to send that over the network.
+   * It signs the messages with the key pair of the user, or the pop token's key pair
+   * according to the type of message.
    *
+   * @param data - The MessageData to be signed and hashed
+   * @param witnessSignatures- The signatures of the witnesses
+   * @returns - The created message
    */
-  public static fromData(
+  public static async fromData(
     data: MessageData, witnessSignatures?: WitnessSignature[],
-  ): Message {
+  ): Promise<Message> {
     const encodedDataJson: Base64UrlData = encodeMessageData(data);
-    const signature: Signature = KeyPairStore.getPrivateKey().sign(encodedDataJson);
+    let publicKey = KeyPairStore.getPublicKey();
+    let privateKey = KeyPairStore.getPrivateKey();
+    let signature: Signature;
 
-    return new Message({
-      data: encodedDataJson,
-      sender: KeyPairStore.getPublicKey(),
-      signature,
-      message_id: Hash.fromStringArray(encodedDataJson.toString(), signature.toString()),
-      witness_signatures: (witnessSignatures === undefined) ? [] : witnessSignatures,
-    });
-  }
-
-  // This function disables the checks of signature and messageID for eleciton result messages
-  // Because the message comes from the back-end and it can't sign the messages since it hasn't
-  // access to the private key
-  // This method is only a temporary solution for the demo and should be removed once a better
-  // solution is found
-  private isElectionResultMessage():boolean {
-    if (this.data.decode().includes('"result":')) {
-      return true;
+    const messageObjectMap = isMessageSignedWithToken.get(data.object);
+    if (!messageObjectMap) {
+      throw new Error(`Message signature for object ${data.object} is unsupported.`);
     }
-    return false;
-  }
-}
 
-/*
-public static async fromData(
-    data: MessageData, witnessSignatures?: WitnessSignature[],
-  ): Promise<Message> {
-    const encodedDataJson: Base64UrlData = encodeMessageData(data);
+    const isSignedWithToken = messageObjectMap.get(data.action);
+    if (isSignedWithToken === undefined) {
```

Comment:
maybe change to !isSignedWithToken so it's more consistent with !messageObjectMap

Reply:
It's a boolean | undefined. I made this mistake first, but then every message signed with the public key threw an error, because the condition was true. It's correct that way